Bluetooth: Controller: ISO-AL fixes for quality tests

niag-Oticon opened the pull request:
- Attempt to use consecutive events (framed TX)
- Correction for Time-offset endianness (framed RX, TX)
-  Validate host timestamps against controller's clock (framed TX)
- Disregard length of PDUs with errors (unframed RX)

**Contains fixes for Ellysis Quality tests related to the ISO-AL:**
- LE/ISO/ISOAL/Req-01 ... 22
- LE/ISO/ISOAL/Rob-01

Fix for local assert related to SDU being larger than expected.

niag-Oticon committed:
Bluetooth: controller: Use consecutive event for framed TX and RX bugfix

Changes:
-- Disregards target event after segmenting the first TX SDU and instead
   decide event based on the next payload and feasibility to fragment
   for that event (time-offset > 0)
-- Corrected error SDU release logic when an error occurs while
   expecting the start of a new SDU, to avoid releasing an SDU which
   could be received in the next payload or event
-- Included additional debug logging

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

niag-Oticon committed:
tests: bluetooth: controller: Test for framed TX in consecutive events

Changes:
-- Included a test to check TX in concesecutive events irrespective of
   the input target event
-- Updated expected output of current tests for change in framed RX
   error handling which waiting for the start of an event

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

niag-Oticon committed:
Bluetooth: controller: corrected time-offset for endianness

Changes:
-- Corrected reading and writing time offset in framed segment header to
   account for endianness
-- Corrected bit positioning of time offset bit field in the segment
   header structure definition
-- Fixed upstream Zephyr BSIM test build failure due to debug logging
   39-bit variable

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

niag-Oticon committed:
test: bluetooth: controller: Updated ISO-AL tests to cover endianness

Changes:
-- Changed RX tests and one TX test to cover endianness of the framed
   segment header fields
-- Updated RX debug PDU print to include framed segment details

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

niag-Oticon committed:
Bluetooth: Controller: ISO-AL validation and selection of TX time stamps

Intent is to pass Ellisys ISOAL quality tests for framed TX scenarios
where the TX SDU includes a time stamp that is not based on the
controller's clock.

Changes:
-- Include controller's reception time as a separate field in the TX SDU
   information
-- Include decision on whether SDU time stamp is valid and based on the
   controller's clock
-- Arbitrate and select / compute time stamp for the SDU

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

niag-Oticon committed:
tests: bluetooth: controller: ISO-AL selection of TX time stamps

Unit tests for TX time stamp related decisions.

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

erbr-ot committed:
Bluetooth: controller: disregard length field on pdu error

In case of unframed iso rx, length field cannot be trusted in case of
PDU error.

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

niag-Oticon committed:
tests: bluetooth: controller: Updated ISO-AL tests for length of err PDU

Updated unit tests to match changes in ISO-AL that disregard the length
of a received PDU with errors.

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

mtpr-ot reviewed (dismissed):
This PR has been tested downstream and eliminates remaining ISO-AL issues in our EBQ/Quality tests (CIS Central/Peripheral). Changes look good.

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -2163,6 +2241,208 @@ static isoal_status_t isoal_update_seg_header_cmplt_length(struct isoal_source *
+				ISOAL_LOG_DBGV("[%p] Using packet_sn for skipped SDUs", source_ctx);
+				sdus_skipped = (tx_sdu->packet_sn - session->last_input_sn) - 1;
+
+			} else if (tx_sdu->packet_sn == session->last_input_sn &&
+					time_diff_valid && time_diff > session->sdu_interval) {
```
```suggestion
			} else if (tx_sdu->packet_sn == session->last_input_sn &&
				   time_diff_valid && time_diff > session->sdu_interval) {
```

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -2163,6 +2241,208 @@ static isoal_status_t isoal_update_seg_header_cmplt_length(struct isoal_source *
+				time_stamp_selected = isoal_get_wrapped_time_us(
+							session->tx_time_stamp,
+							time_diff - session->tx_time_offset);
+				ISOAL_LOG_DBGV("[%p] Projecting Time Stamp (%lu) from SDU delta",
+						source_ctx, time_stamp_selected);
```
```suggestion
				ISOAL_LOG_DBGV("[%p] Projecting Time Stamp (%lu) from SDU delta",
					       source_ctx, time_stamp_selected);
```

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -2163,6 +2241,208 @@ static isoal_status_t isoal_update_seg_header_cmplt_length(struct isoal_source *
+				 * calcutation
+				 */
+				time_stamp_selected = tx_sdu->time_stamp;
+				ISOAL_LOG_DBGV("[%p] Selecting Time Stamp (%lu) from SDU",
+						source_ctx, time_stamp_selected);
```
```suggestion
				ISOAL_LOG_DBGV("[%p] Selecting Time Stamp (%lu) from SDU",
					       source_ctx, time_stamp_selected);
```

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -2163,6 +2241,208 @@ static isoal_status_t isoal_update_seg_header_cmplt_length(struct isoal_source *
+		 * released for this event. Segmentation should continue from
+		 * that payload.
+		 */
+		next_payload_number = MAX(pp->payload_number,
+					 (actual_event * session->burst_number));
```
```suggestion
		next_payload_number = MAX(pp->payload_number,
					  (actual_event * session->burst_number));
```

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -2163,6 +2241,208 @@ static isoal_status_t isoal_update_seg_header_cmplt_length(struct isoal_source *
+							session->iso_interval * ISO_INT_UNIT_US);
+		}
+
+		ISOAL_LOG_DBGV("[%p] Chosen PL=%llu Evt=%llu Ref=%lu",
+				source_ctx,  (actual_event * session->burst_number), actual_event,
+				actual_grp_ref_point);
```
```suggestion
		ISOAL_LOG_DBGV("[%p] Chosen PL=%llu Evt=%llu Ref=%lu",
			       source_ctx, (actual_event * session->burst_number), actual_event,
			       actual_grp_ref_point);
```

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -2163,6 +2241,208 @@ static isoal_status_t isoal_update_seg_header_cmplt_length(struct isoal_source *
+		next_payload_number = MAX(pp->payload_number,
+					 (actual_event * session->burst_number));
+
+	ISOAL_LOG_DBGV("[%p] Final Evt=%llu (PL=%llu) Ref.=%lu Next PL=%llu",
+			source, actual_event, (actual_event * session->burst_number),
+			actual_grp_ref_point, next_payload_number);
```
```suggestion
		ISOAL_LOG_DBGV("[%p] Final Evt=%llu (PL=%llu) Ref.=%lu Next PL=%llu",
			       source, actual_event, (actual_event * session->burst_number),
			       actual_grp_ref_point, next_payload_number);
```

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -2163,6 +2241,208 @@ static isoal_status_t isoal_update_seg_header_cmplt_length(struct isoal_source *
+		time_diff_valid = isoal_get_time_diff(time_stamp_selected,
+					actual_grp_ref_point, &time_diff);
+
+		LL_ASSERT(time_diff_valid);
+		LL_ASSERT(time_diff > 0);
+		/* Time difference must be less than the maximum possible
+		 * time-offset of 24-bits.
+		 */
+		LL_ASSERT(time_diff <= 0x00FFFFFF);
```
Wouldn't `LL_ASSERT(time_diff_valid);` be enough?

niag-Oticon replied:
No the time difference is valid even if it is 0 and the specification says the time offset should be a positive value. I'm not sure if 0 should be considered a positive value, but if the time offset is 0 I don't think it will make it in time for the air anyway. I am considering importing a vendor specific profiled value in deciding margins, but that's for another PR.

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -2202,15 +2481,58 @@ static isoal_status_t isoal_tx_framed_produce(isoal_source_handle_t source_hdl,
 	zero_length_sdu = (packet_available == 0 &&
 		tx_sdu->sdu_state == BT_ISO_SINGLE);
 
+	ISOAL_LOG_DBGV("[%p] SDU %u len=%u TS=%lu Ref=%lu Evt=%llu Frag=%u",
+		source, tx_sdu->packet_sn, tx_sdu->iso_sdu_length, tx_sdu->time_stamp,
+		tx_sdu->grp_ref_point, tx_sdu->target_event, tx_sdu->sdu_state);
```
```suggestion
	ISOAL_LOG_DBGV("[%p] SDU %u len=%u TS=%lu Ref=%lu Evt=%llu Frag=%u",
		       source, tx_sdu->packet_sn, tx_sdu->iso_sdu_length, tx_sdu->time_stamp,
		       tx_sdu->grp_ref_point, tx_sdu->target_event, tx_sdu->sdu_state);
```

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -2224,74 +2546,34 @@ static isoal_status_t isoal_tx_framed_produce(isoal_source_handle_t source_hdl,
+		/* Update input packet number and time stamp */
+		session->last_input_sn = tx_sdu->packet_sn;
+
+		if (pp->initialized && tx_sdu->time_stamp == tx_sdu->cntr_time_stamp &&
+			(!time_diff_valid || time_diff < session->sdu_interval)) {
```
```suggestion
		if (pp->initialized && tx_sdu->time_stamp == tx_sdu->cntr_time_stamp &&
		    (!time_diff_valid || time_diff < session->sdu_interval)) {
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_common.c:
```diff
@@ -122,12 +123,13 @@ uint16_t isoal_test_insert_segment(bool sc, bool cmplt, uint32_t time_offset, ui
-	seg_hdr.sc = sc;
-	seg_hdr.cmplt = cmplt;
-	seg_hdr.len = length + (sc ? 0 : PDU_ISO_SEG_TIMEOFFSET_SIZE);
+	/* Check endianness */
```
It is not clear to me what this comment refers to

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_debug.c:
```diff
@@ -60,8 +66,39 @@ void isoal_test_debug_print_rx_pdu(struct isoal_pdu_rx *pdu_meta)
 
-	for (int i = 0; i < pdu_meta->pdu->len; i++) {
-		PRINT("%02x ", pdu_meta->pdu->payload[i]);
+	for (int i = 0; i < pdu->len; i++) {
```
```suggestion
	for (uint8_t i = 0; i < pdu->len; i++) {
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_debug.c:
```diff
@@ -60,8 +66,39 @@ void isoal_test_debug_print_rx_pdu(struct isoal_pdu_rx *pdu_meta)
-	for (int i = 0; i < pdu_meta->pdu->len; i++) {
-		PRINT("%02x ", pdu_meta->pdu->payload[i]);
+	for (int i = 0; i < pdu->len; i++) {
+		if (seg_length == 0 && pdu->ll_id == PDU_BIS_LLID_FRAMED) {
```
```suggestion
		if (seg_length == 0U && pdu->ll_id == PDU_BIS_LLID_FRAMED) {
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_debug.c:
```diff
@@ -60,8 +66,39 @@ void isoal_test_debug_print_rx_pdu(struct isoal_pdu_rx *pdu_meta)
-		PRINT("%02x ", pdu_meta->pdu->payload[i]);
+	for (int i = 0; i < pdu->len; i++) {
+		if (seg_length == 0 && pdu->ll_id == PDU_BIS_LLID_FRAMED) {
+			seg_length = pdu->payload[i+1];
```
```suggestion
			seg_length = pdu->payload[i + 1U];
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_debug.c:
```diff
@@ -60,8 +66,39 @@ void isoal_test_debug_print_rx_pdu(struct isoal_pdu_rx *pdu_meta)
+			PRINT("[%s %s %03u]",
+				pdu->payload[i] & BIT(0) ? "C" : "S",
+				pdu->payload[i] & BIT(1) ? "C" : "-",
+				pdu->payload[i+1]);
```
```suggestion
				pdu->payload[i + 1U]);
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_debug.c:
```diff
@@ -60,8 +66,39 @@ void isoal_test_debug_print_rx_pdu(struct isoal_pdu_rx *pdu_meta)
+				pdu->payload[i] & BIT(0) ? "C" : "S",
+				pdu->payload[i] & BIT(1) ? "C" : "-",
+				pdu->payload[i+1]);
+			if ((pdu->payload[i] & BIT(0)) == 0) {
```
```suggestion
			if ((pdu->payload[i] & BIT(0)) == 0U) {
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_debug.c:
```diff
@@ -60,8 +66,39 @@ void isoal_test_debug_print_rx_pdu(struct isoal_pdu_rx *pdu_meta)
+				pdu->payload[i+1]);
+			if ((pdu->payload[i] & BIT(0)) == 0) {
+				PRINT("(%8uus)",
+					((uint32_t)pdu->payload[i+2] +
+					((uint32_t)pdu->payload[i+3] << 8) +
+					((uint32_t)pdu->payload[i+4] << 16)));
```
```suggestion
					((uint32_t)pdu->payload[i + 2U] +
					((uint32_t)pdu->payload[i + 3U] << 8) +
					((uint32_t)pdu->payload[i + 4U] << 16)));
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_debug.c:
```diff
@@ -60,8 +66,39 @@ void isoal_test_debug_print_rx_pdu(struct isoal_pdu_rx *pdu_meta)
+			PRINT(" / ");
+			PRINT("[%02x %02x]",
+				pdu->payload[i],
+				pdu->payload[i+1]);
```
```suggestion
				pdu->payload[i + 1U]);
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_debug.c:
```diff
@@ -60,8 +66,39 @@ void isoal_test_debug_print_rx_pdu(struct isoal_pdu_rx *pdu_meta)
+			PRINT("[%02x %02x]",
+				pdu->payload[i],
+				pdu->payload[i+1]);
+			if ((pdu->payload[i] & BIT(0)) == 0) {
```
```suggestion
			if ((pdu->payload[i] & BIT(0)) == 0U) {
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_debug.c:
```diff
@@ -60,8 +66,39 @@ void isoal_test_debug_print_rx_pdu(struct isoal_pdu_rx *pdu_meta)
+				pdu->payload[i+1]);
+			if ((pdu->payload[i] & BIT(0)) == 0) {
+				PRINT("(%02x %02x %02x)",
+					(uint32_t)pdu->payload[i+4],
+					(uint32_t)pdu->payload[i+3],
+					(uint32_t)pdu->payload[i+2]);
```
```suggestion
					(uint32_t)pdu->payload[i + 4U],
					(uint32_t)pdu->payload[i + 3U],
					(uint32_t)pdu->payload[i + 2U]);
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_debug.c:
```diff
@@ -60,8 +66,39 @@ void isoal_test_debug_print_rx_pdu(struct isoal_pdu_rx *pdu_meta)
+			}
+			PRINT(" : ");
+			seg_length -= pdu->payload[i] & BIT(0) ? 0 : PDU_ISO_SEG_TIMEOFFSET_SIZE;
+			i = i + PDU_ISO_SEG_HDR_SIZE +
```
```suggestion
			i += PDU_ISO_SEG_HDR_SIZE +
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_debug.c:
```diff
@@ -60,8 +66,39 @@ void isoal_test_debug_print_rx_pdu(struct isoal_pdu_rx *pdu_meta)
+			seg_length -= pdu->payload[i] & BIT(0) ? 0 : PDU_ISO_SEG_TIMEOFFSET_SIZE;
+			i = i + PDU_ISO_SEG_HDR_SIZE +
+				(pdu->payload[i] & BIT(0) ? 0 : PDU_ISO_SEG_TIMEOFFSET_SIZE);
+		}
+		PRINT("%02x ", pdu->payload[i]);
```
```suggestion
		}
		
		PRINT("%02x ", pdu->payload[i]);
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_debug.c:
```diff
@@ -60,8 +66,39 @@ void isoal_test_debug_print_rx_pdu(struct isoal_pdu_rx *pdu_meta)
+					((uint32_t)pdu->payload[i+2] +
+					((uint32_t)pdu->payload[i+3] << 8) +
+					((uint32_t)pdu->payload[i+4] << 16)));
+			}
+			PRINT(" / ");
```
```suggestion
			}
			
			PRINT(" / ");
```

Thalley reviewed (changes requested):
Mostly code style comments

niag-Oticon commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -46,10 +48,18 @@ LOG_MODULE_REGISTER(bt_ctlr_isoal, LOG_LEVEL_INF);
 
 #include "hal/debug.h"
 
+#define TRUE         1U
+#define FALSE        0U
```
I am defining these because there are bit fields used as boolean packed into the unused bits of a uint64 that only uses 39 bits. The code is more readable if the assignment is either true or false. As I understand it using the bool type definitions with the bit fields makes our compiler unhappy.

I didn't see common definitions that I could reuse but if there is a standard approach that I missed, please let me know.

Thalley replied:
That's also what I thought. I do, however, think that this should be a generic approach for the controller, and I would suggest to (if you really want to add this) move this to a new PR that defines these a common place in the controller (if we don't already have anything like this), and apply it everywhere in the controller where we use a bit in a bitfield as a boolean value. 

I do not think this should be defined in a single .c file and only applied there

niag-Oticon replied:
Created a new PR https://github.com/zephyrproject-rtos/zephyr/pull/61591 for the definitions.

niag-Oticon replied:
The conclusion reached in the controller meeting today was that definition of `TRUE` and `FALSE` was unnecessary. So as agreed I have reverted to use 1U and 0U instead.

niag-Oticon commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -1806,15 +1877,26 @@ uint16_t isoal_tx_unframed_get_next_payload_number(isoal_source_handle_t source_
+ * receiver for each SDU and there were problems  with LL/CIS/PER/BV-39-C which
  * expects clustering within an event.
  *
+ * @par
```
Yes but this is an extension of the `note` so as I understand it will end on a blank line unless extended by the `par`. I was following the example for multi paragraph [note](https://www.doxygen.nl/manual/commands.html#cmdpar).

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -1566,6 +1591,46 @@ void isoal_source_destroy(isoal_source_handle_t hdl)
+				      const uint32_t cntr_time,
+				      const uint32_t time_stamp)
+{
+	struct isoal_source_session *session;
```
```suggestion
	const struct isoal_source_session *session;
```

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -1566,6 +1591,46 @@ void isoal_source_destroy(isoal_source_handle_t hdl)
 	isoal_source_deallocate(hdl);
 }
 
+static bool isoal_is_time_stamp_valid(struct isoal_source *source_ctx,
```
```suggestion
static bool isoal_is_time_stamp_valid(const struct isoal_source *source_ctx,
```

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -1610,6 +1675,10 @@ static isoal_status_t isoal_tx_pdu_emit(const struct isoal_source *source_ctx,
 	/* Attempt to enqueue the node towards the LL */
 	status = source_ctx->session.pdu_emit(node_tx, handle);
 
+	ISOAL_LOG_DBG("[%p] PDU %llu err=%X len=%u frags=%u released",
+		      source_ctx, payload_number, status,
+		      produced_pdu->contents.pdu->len, sdu_fragments);
```
Outside the scope of the PR, but what's the deal with `ISOAL_LOG_DBG`? It's defined as
```c
#define ISOAL_LOG_DBG(...)     LOG_DBG(__VA_ARGS__)

#if defined(ISOAL_DEBUG_VERBOSE)
#define ISOAL_LOG_DBGV(...)    LOG_DBG(__VA_ARGS__)
#else
#define ISOAL_LOG_DBGV(...)    (void) 0
#endif /* ISOAL_DEBUG_VERBOSE */
```

So it does nothing unless `ISOAL_DEBUG_VERBOSE` is set, but why not just use it `LOG_DBG`?

niag-Oticon replied:
I'm not certain if I understood the comment correctly, but just to point out that there are two macros there. `ISOAL_LOG_DBG()` and `ISOAL_LOG_DBGV()` with the latter being the verbose version. The standard logging does not provide tiers to control verbosity on debug statements. The verbose statements are useful when debugging the unit tests, but probably a disruption if logging is enable elsewhere.

Thalley replied:
Ah, I misread; didn't see the `V` in `ISOAL_LOG_DBGV`. Arguably that should have been called `ISOAL_LOG_DBG_VERBOSE` to avoid the confusion (i.e. ironically make the macro more verbose). 

Have you considered moving `ISOAL_DEBUG_VERBOSE` to a Kconfig option, similar to https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/Kconfig#L952 ?

niag-Oticon replied:
I did consider a longer macro name, but ideally I would have liked these statements to be one liners. A longer macro name means less space for the statement. If there are strong objections let me know otherwise I would like to keep it short.

Thanks for the example kconfig. Let me consider that.

niag-Oticon replied:
Done now.

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -2163,6 +2241,209 @@ static isoal_status_t isoal_update_seg_header_cmplt_length(struct isoal_source *
+ * @param[out] time_offset    Segmentation Time offset to selected event
+ * @return                The number SDUs skipped from the last
+ */
+static uint16_t isoal_tx_framed_find_correct_tx_event(struct isoal_source *source_ctx,
+						      const struct isoal_sdu_tx *tx_sdu,
+						      uint64_t *payload_number,
+						      uint32_t *grp_ref_point,
+						      uint32_t *time_offset)
+{
+	struct isoal_source_session *session;
+	struct isoal_pdu_production *pp;
```
```suggestion
static uint16_t isoal_tx_framed_find_correct_tx_event(const struct isoal_source *source_ctx,
						      const struct isoal_sdu_tx *tx_sdu,
						      uint64_t *payload_number,
						      uint32_t *grp_ref_point,
						      uint32_t *time_offset)
{
	const struct isoal_source_session *session;
	const struct isoal_pdu_production *pp;
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_debug.c:
```diff
@@ -60,8 +66,41 @@ void isoal_test_debug_print_rx_pdu(struct isoal_pdu_rx *pdu_meta)
 
-	for (int i = 0; i < pdu_meta->pdu->len; i++) {
-		PRINT("%02x ", pdu_meta->pdu->payload[i]);
+	for (uint8_t i = 0; i < pdu->len; i++) {
```
```suggestion
	for (uint8_t i = 0U; i < pdu->len; i++) {
```

Thalley commented on tests/bluetooth/ctrl_isoal/src/isoal_test_debug.c:
```diff
@@ -60,8 +66,41 @@ void isoal_test_debug_print_rx_pdu(struct isoal_pdu_rx *pdu_meta)
+					(uint32_t)pdu->payload[i + 4U],
+					(uint32_t)pdu->payload[i + 3U],
+					(uint32_t)pdu->payload[i + 2U]);
+			}
+			PRINT(" : ");
```
```suggestion
			}
			
			PRINT(" : ");
```

Thalley reviewed (dismissed):
All my comments have been resolved

niag-Oticon committed:
Bluetooth: controller: Included kconfigs for ISO-AL logging

Included kconfigs to set the ISO-AL logging level and control debug
logging verbosity.

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

niag-Oticon committed:
tests: bluetooth: controller: Included kconfigs for ISO-AL logging

Included kconfigs to set the ISO-AL logging level and control debug
logging verbosity in ISO-AL unit tests.

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

Thalley commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -206,6 +206,21 @@ config BT_CTLR_ISO_TX_BUFFER_SIZE
 	  Size of the Isochronous Tx buffers and the value returned in HCI LE
 	  Read Buffer Size V2 command response.
 
+config BT_CTLR_ISOAL_LOG_LEVEL
+	int "ISO-AL logging level"
+	range 0 4
+	default 3
+	help
+	  Use this option to enable ISO-AL logging at the given level.
+	  Default is LOG_LEVEL_INF (3).
```
Instead of doing this, you should use the template for modules as per https://docs.zephyrproject.org/latest/services/logging/index.html#logging-in-a-module

So it would be something like 

```c
module = BT_CTLR_ISOAL
module-str = "ISO-AL"
source "subsys/logging/Kconfig.template.log_config"
```

niag-Oticon replied:
Should be done now. Not exactly the same because it looked as if there was need to support legacy logging as well.

Thalley commented on tests/bluetooth/ctrl_isoal/Kconfig:
```diff
@@ -12,6 +12,21 @@ config BT_CTLR_CONN_ISO_GROUPS
 	range 1 240
 	default 1
 
+config BT_CTLR_ISOAL_LOG_LEVEL
+	int "ISO-AL logging level"
+	range 0 4
+	default 3
+	help
+	  Use this option to enable ISO-AL logging at the given level.
+	  Default is LOG_LEVEL_INF (3).
+
+config BT_CTLR_ISOAL_LOG_DBG_VERBOSE
+	bool "ISO-AL verbose debug logging"
+	depends on BT_CTLR_ISOAL_LOG_LEVEL = 4
+	default n
+	help
+	  Use this option to enable ISO-AL verbose debug logging.
```
Why do we need to define this in two places?

niag-Oticon replied:
Some of the configurations used in the ISO-AL unit tests have a long chain of dependencies. Redefining it is easier for the unit test configuration rather than trying figure out the chain and trying to fulfill the whole chain.

mtpr-ot reviewed (approved):
LGTM
